### DISCUSSION TITLE
Gnmt train slow

### DISCUSSION THREAD

**Discussion by yajiedesign** [dmlc/gluon-nlp#668] [Q&A]
GPU cannot be filled with train_gnmt.
GPU:GTX TITAN X
CPU:4770k
GPU LOAD: about 10%
CPU LOAD: one core full.
dataset: LCSTS 
vocab size: 320783
num work: 0 (Trying to set it to 2 seems to improve, but it takes up too much memory (about six times as much as 0)).
batch size: 16

**Comment by ymjiang** [comment]
Batch size=16 seems not reasonable. Did you try to increase it?

**Comment by yajiedesign** [comment]
@ymjiang There's no way to increase it any more. Now it takes up about 10G of gpu memory.

**Comment by szha** [comment]
cc @sxjscience. One way to work around the memory limit is to perform gradient accumulation. Several model zoo scripts already implement this logic, and it might be a good idea to add it to GNMT too.

**Comment by sxjscience** [comment]
Will take a look at it.

**Comment by szhengac** [comment]
@zhreshold Using `num_worker=2` requires six times more memory compared to `num_worker=0` seems unreasonable. Do you have any idea?

**Comment by zhreshold** [comment]
@szhengac using multi worker will enable pre-fetching batches, default is 2*batch size, and each worker will be fetching a batch simultaneously.

**Comment by yajiedesign** [comment]
now.I saved gpu money by setting share_embed.and I increase batch size.it some improvements.

**Comment by szhengac** [comment]
@zhreshold but each input batch is simply an integer matrix, which shouldn't occupy too much gpu memory.

**Comment by zhreshold** [comment]
@szhengac Did you mean gpu memory? Dataloader should never occupy gpu memory

**Comment by szhengac** [comment]
@zhreshold Oh I mean cpu memory.
